lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py

```diff
@@ -0,0 +1,308 @@
+      aliases: ['datastore_name']
+      required: yes
+      type: str
+    source:
+      description:
+      - The full path of the local file to push to vCenter.
+      required: yes
+      type: str
+    destination:
+      description:
+      - The full path to the destination file in the datastore.
+      required: yes
+      type: str
+    debounce:
+      description:
+        - Uploading large files unnecessarily can be avoided
+          with the C(debounce) option.
+        - C(hash) will hash C(source) (using SHA256) and check
+          for a file named C(<destination>.sha256) - if that file
+          does not exist, or the hash in that file differs from
+          the local hash, the file is uploaded and the hash is
+          placed into C(<destination>.sha256) to debounce the next
+          upload.  This option requires no write access locally.
+        - C(hashcache) is the same as C(hash) except that it requires
+          local write access to C(<source>.sha256) so the hash can
+          be cached locally.  If C(source) is modified then the hash
+          will be recalculated on the next upload.  This option
+          prevents the local server from re-hashing large data files.
+        - If C(none), the file will always be uploaded and no
+          C(.sha256) files will be used.  This is appropriate
+          for smaller files.
+      default: none
+      choices: [ none, hash, hashcache ]
+      type: str
+extends_documentation_fragment: vmware.documentation
+'''
+
+EXAMPLES = '''
+- name: Upload, overwriting any existing file at the destination.
+  vmware_datastore_upload:
+    hostname: "10.0.1.1"
+    username: "myuser"
+    password: "mypass"
+    datastore: "Dev"
+    source: "/data/linux.iso"
+    destination: "/ISO/linux.iso"
+'''
+
+RETURN = '''
+results:
+    description: result of the operation
+    returned: always
+    type: dict
+    sample:
+      "changed": true
+      "msg": "File '/data/linux.iso' uploaded to '[Dev] ISO/linux.iso' through host 10.0.1.100."
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_native
+from ansible.module_utils.urls import fetch_url
+from ansible.module_utils.vmware import PyVmomi, vmware_argument_spec
+
+import ansible.module_utils.six.moves.urllib.parse as urllib_parse
+import atexit
+import hashlib
+import mmap
+import os
+
+
+class VmwareDatastoreUploadMgr(PyVmomi):
+    def __init__(self, module):
+        ''' Prepares for the upload. '''
+        super(VmwareDatastoreUploadMgr, self).__init__(module)
+        datacenter_name = self.params.get('datacenter')
+        datastore_name = self.params.get('datastore')
+        self._datacenter = None
+        if datacenter_name:
+            self._datacenter = self.find_datacenter_by_name(
+                datacenter_name=datacenter_name)
+        self._datastore = self.find_datastore_by_name(
+            datastore_name, datacenter=self.datacenter)
```

Comment:
Seems like a bug? `self.datacenter` doesn't exist
```suggestion
            datastore_name, datacenter=self._datacenter)
```